Mathlib/AlgebraicGeometry/Noetherian.lean

```diff
@@ -0,0 +1,352 @@
+  quasi-separated.
+
+* `AlgebraicGeometry.isNoetherian_iff_of_finite_affine_openCover`: A scheme is Noetherian
+  if and only if it is covered by finitely many affine opens whose sections are Noetherian rings.
+
+* `AlgebraicGeometry.IsNoetherian.noetherianSpace`: A Noetherian scheme is
+  topologically a Noetherian space.
+
+## References
+
+* [Stacks: Noetherian Schemes](https://stacks.math.columbia.edu/tag/01OU)
+* [Robin Hartshorne, *Algebraic Geometry*][Har77]
+
+-/
+
+universe u v
+
+open Opposite AlgebraicGeometry Localization IsLocalization TopologicalSpace
+
+namespace AlgebraicGeometry
+
+/-- A scheme `X` is locally Noetherian if `𝒪ₓ(U)` is Noetherian for all affine `U`. -/
+class IsLocallyNoetherian (X : Scheme) : Prop where
+  component_noetherian : ∀ (U : X.affineOpens),
+    IsNoetherianRing Γ(X, U) := by infer_instance
+
+section localizationProps
+
+variable {R : Type u} [CommRing R] (S : Finset R) (hS : Ideal.span (α := R) S = ⊤)
+  (hN : ∀ s : S, IsNoetherianRing (Away (M := R) s))
+
+/-- Let `R` be a ring, and `f i` a finite collection of elements of `R` generating the unit ideal.
+If the localization of `R` at each `f i` is noetherian, so is `R`.
+
+We follow the proof given in [Har77], Proposition II.3.2 -/
+theorem isNoetherianRing_of_away : IsNoetherianRing R := by
+  apply Iff.mp
+  apply monotone_stabilizes_iff_noetherian
+  intro I
+  let floc s := algebraMap R (Away (M := R) s)
+  let suitableN s :=
+    { n : ℕ | ∀ m : ℕ, n ≤ m → (Ideal.map (floc s) (I n)) = (Ideal.map (floc s) (I m)) }
+  let minN s := sInf (suitableN s)
+  have hSuit : ∀ s : S, minN s ∈ suitableN s := by
+    intro s
+    apply Nat.sInf_mem
+    let f : ℕ →o Ideal (Away (M := R) s) :=
+      ⟨fun n => Ideal.map (floc s) (I n), by
+        intro n m hnm
+        dsimp
+        apply Ideal.map_mono
+        exact I.monotone hnm⟩
+    exact monotone_stabilizes_iff_noetherian.mpr (hN s) f
+  let N := Finset.sup S minN
+  use N
+  have hN : ∀ s : S, minN s ≤ N := fun s => Finset.le_sup s.prop
+  intro n hn
+  rw [IsLocalization.ideal_eq_iInf_comap_map_away hS (I N),
+      IsLocalization.ideal_eq_iInf_comap_map_away hS (I n),
+      iInf_subtype', iInf_subtype']
+  apply iInf_congr
+  intro s
+  congr 1
+  rw [← hSuit s N (hN s)]
+  exact hSuit s n <| Nat.le_trans (hN s) hn
+
+end localizationProps
+
+variable {X : Scheme}
+
+/-- If a scheme `X` has a cover by affine opens whose sections are Noetherian rings,
+then `X` is locally Noetherian. -/
+theorem isLocallyNoetherian_of_affine_cover {S : Set X.affineOpens}
+    (hS : (⨆ i ∈ S, i : Opens X) = ⊤)
+    (hS' : ∀ U ∈ S, IsNoetherianRing Γ(X, U)) : IsLocallyNoetherian X := by
+  refine ⟨fun U => ?_⟩
+  rw [iSup_subtype'] at hS
+  apply of_affine_open_cover (P := _) (fun i : S => i) hS U _ _ (fun U => hS' U.1 U.2)
+  · intro U f hN
+    have := U.prop.isLocalization_basicOpen f
+    exact IsLocalization.isNoetherianRing (.powers f) Γ(X, X.basicOpen f) hN
+  · intro U s _ hN
+    let R := Γ(X, U)
```

Comment:
It seems that `R` is used only once, so that it is probably not necessary to declare it separately.